### ISSUE TITLE
bridge2: O(n^2) compute and bandwidth usage for session

### ISSUE THREAD

**Issue by ajbouh** [ajbouh/substrate#180]
We are currently snapshotting, sorting, encoding, sending, receiving, decoding the entire session on every session event for every listener. This leads to O(n^2) bandwidth and compute requirements.

We should switch to only snapshotting when a new listener arrives and sending events as they happen.

**Comment by mgood** [comment]
yeah, there's some discussion about sending events instead of full updates in #85 as well, so I'll link these together
